File: app/src/main/java/workshop1024/com/xproject/controller/adapter/PublisherListAdapter.java
package workshop1024.com.xproject.controller.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.model.publisher.Publisher;

/**
 * 发布者列表选择适配器
 */
public class PublisherListAdapter extends RecyclerView.Adapter<PublisherListAdapter.PublisherViewHolder> {
    private List<Publisher> mPublisherList;
    private OnPublisherListSelectListener mOnPublisherListSelectListener;

    public PublisherListAdapter(List<Publisher> publisherList, OnPublisherListSelectListener
            onPublisherListSelectListener) {
        mPublisherList = publisherList;
        mOnPublisherListSelectListener = onPublisherListSelectListener;
    }

    @Override
    public PublisherViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.publishlist_item_content, parent,
                false);
        return new PublisherViewHolder(view, mOnPublisherListSelectListener);
    }

    @Override
    public void onBindViewHolder(final PublisherViewHolder holder, int position) {
        Publisher publisher = mPublisherList.get(position);

        holder.mNameTextView.setText(publisher.getName());
        holder.mSubscribeNumTextView.setText(publisher.getSubscribeNum());
        holder.mSelectedCheckBox.setChecked(publisher.isIsSubscribed());
        holder.mPublisher = publisher;
    }

    @Override
    public int getItemCount() {
        return mPublisherList.size();
    }

    public interface OnPublisherListSelectListener {
        void publisherListItemSelect(Publisher selectPublisher, boolean isSelected);
    }

    public class PublisherViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public final ImageView mIconImageView;
        public final TextView mNameTextView;
        private final TextView mSubscribeNumTextView;
        private final CheckBox mSelectedCheckBox;
        private Publisher mPublisher;
        private OnPublisherListSelectListener mOnPublisherListSelectListener;

        public PublisherViewHolder(View view, OnPublisherListSelectListener onPublisherListSelectListener) {
            super(view);
            mOnPublisherListSelectListener = onPublisherListSelectListener;

            mIconImageView = view.findViewById(R.id.publisherlist_imageview_icon);
            mNameTextView = view.findViewById(R.id.publisherlist_textview_name);
            mSubscribeNumTextView = view.findViewById(R.id.publisherlist_textview_subscribenum);
            mSelectedCheckBox = view.findViewById(R.id.publisherlist_checkbox_selected);

            mSelectedCheckBox.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            mOnPublisherListSelectListener.publisherListItemSelect(mPublisher, mSelectedCheckBox.isChecked());
        }
    }
}

File: main/src/main/java/workshop1024/com/xproject/main/controller/adapter/PublisherListAdapter.kt
package workshop1024.com.xproject.main.controller.adapter

import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import workshop1024.com.xproject.main.R
import workshop1024.com.xproject.main.databinding.PublishlistItemContentBinding

/**
 * 发布者列表选择适配器
 */
class PublisherListAdapter(private val mPublisherList: List<workshop1024.com.xproject.main.model.publisher.Publisher>, private val mOnPublisherListSelectListener: OnPublisherListSelectListener?)
    : RecyclerView.Adapter<PublisherListAdapter.PublisherViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PublisherViewHolder {
        val publishlistItemContentBinding = DataBindingUtil.inflate<PublishlistItemContentBinding>(LayoutInflater.from(parent.context),
                R.layout.publishlist_item_content, parent, false)
        publishlistItemContentBinding.publisherHandlers = PublisherHandlers()
        return PublisherViewHolder(publishlistItemContentBinding)
    }

    override fun onBindViewHolder(holder: PublisherViewHolder, position: Int) {
        holder.mPublishlistItemContentBinding.publisher = mPublisherList[position]
    }

    override fun getItemCount(): Int {
        return mPublisherList.size
    }

    interface OnPublisherListSelectListener {
        fun onPublisherListItemSelect(selectPublisher: workshop1024.com.xproject.main.model.publisher.Publisher, isSelected: Boolean)
    }

    inner class PublisherViewHolder(internal val mPublishlistItemContentBinding: PublishlistItemContentBinding) : RecyclerView.ViewHolder(mPublishlistItemContentBinding.root)

    inner class PublisherHandlers {
        fun onTouchSelected(v: View, event: MotionEvent, publisher: workshop1024.com.xproject.main.model.publisher.Publisher): Boolean {
            //CheckBox选中后，先不更改状态，待请求结果后在更改状态，参考 https://blog.csdn.net/qq_37822393/article/details/80195090
            if (event.action == MotionEvent.ACTION_UP) {
                Log.i("XProject", "public boolean onTouchSelected, publisher = " + publisher.typeId)
                mOnPublisherListSelectListener?.onPublisherListItemSelect(publisher, !(v as CheckBox).isChecked)
            }
            return true
        }
    }
}